starknet_devnet/predeployed_contract_wrapper.py

```diff
@@ -1,9 +1,100 @@
 """Parent for predeployed contract wrapper classes"""
+from abc import ABC, abstractmethod
 
+from starkware.starknet.testing.starknet import Starknet
+from starkware.starknet.services.api.contract_class import ContractClass
+from starkware.starknet.testing.contract import StarknetContract
 
-class PredeployedContractWrapper:
+
+class PredeployedContractWrapper(ABC):
     """Parent class for all predeployed contract wrapper classes"""
 
+    # Cannot import it because of circular imports
+    # from .starknet_wrapper import StarknetWrapper
+
+    starknet_wrapper = NotImplemented
+
+    @property
+    def address(self) -> int:
+        try:
+            return self.ADDRESS
+        except AttributeError:
+            try:
+                return self._address
+            except AttributeError:
+                raise NotImplementedError(
+                    "Predeployed contracts should provide an address attribute or an ADDRESS class attribute"
+                )
+
+    @address.setter
+    def address(self, address: int):
+        self._address = address
+
+    @property
+    def class_hash_bytes(self) -> bytes:
+        try:
+            return self._class_hash_bytes
+        except AttributeError:
+            try:
+                return self.HASH_BYTES
+            except AttributeError:
+                raise NotImplementedError(
+                    "Predeployed contracts should provide a class_hash_bytes attribute or a HASH_BYTES class attribute"
+                )
+
+    @class_hash_bytes.setter
+    def class_hash_bytes(self, class_hash_bytes: bytes):
```

Comment:
I don't understand the need for this? Is this approach a Python idiom?

Reply:
We need the setter to be able to set the value inside `__init__` with `self.class_hash_bytes = ...` because `class_hash_bytes` is a `property`

This way, classes can either provide a `HASH_BYTES` or set the value directly inside `__main__` or even override the `property` altogether

Reply:
So, what do we do with this ?

Reply:
I refuse to believe that nested try-except blocks are a good approach. There is a global function called `hasattr`, but I would avoid even using that.

The base class could specify unimplemented (abstract) getters, and child classes could implement them by returning the precalculated values that are stored as their class properties (e.g. `HASH`). I haven't tried coding this proposal, so I don't know how it looks, but it sounds better to me.